lib/utils/report_output.rb

```diff
@@ -0,0 +1,63 @@
+require "securerandom"
+
+# An attempt at simplifying the structure of report output dirs and files.
+# Give report_name as input.
+# The methods and their return:
+#   id: the "unique-ish" part of the filename
+#   dir: path to an output dir
+#   file: path to an output file
+#   handle("r"): an open read handle for file
+#   handle("w"): an open write handle for file
+# Both dir and file will contain the report name,
+# and the file a timestamp and a unique-ish string.
+# dir is autovivified with mkdir_p.
+# Example:
+# output = Utils::ReportOutput.new("foo", ".txt")
+# output.dir # -> "/local_reports/foo"
+# output.file # -> "/local_reports/foo/foo_YYYYMMDD_abcd1234.txt"
+
+module Utils
+  class ReportOutput
+    attr_reader :report_name, :ext
+    def initialize(report_name, ext = ".tsv")
+      @report_name = report_name # e.g. "cost_report" or "cost_report_umich"
+      @ext = ext
+    end
+
+    def file
+      if @file.nil?
+        @file = File.join(dir, id) + ext
+      end
+      @file
+    end
+
+    def dir(mkdir: true)
+      # In testing it is useful to get the path without mkdir_p,
+      # hence the mkdir param
+      dir_path = File.join(Settings.local_report_path || "/tmp", report_name)
+      if @dir.nil?
+        @dir = dir_path
+      end
+      if mkdir
+        FileUtils.mkdir_p(dir_path)
+      end
+      @dir
+    end
+
+    def id
+      if @id.nil?
+        timestamp = Time.now.strftime("%Y%m%d_%H%M%S")
+        # Adding a random string in case 2 identical jobs that write a file
```

Comment:
How likely is this to happen in practice? Will the extra hex digits cause any extra work or confusion when sending the reports to others?

Reply:
I have for sure already been bitten by this, where I have 2 competing (and possibly pathological) sidekiq jobs that are writing to the same file. Is it a bit of belt and suspenders? Perhaps. I'm keeping it until such a time when we start regretting it, easy enough to undo.